src/librbd/io/ImageRequest.cc

```diff
@@ -826,6 +834,87 @@ void ImageWriteSameRequest<I>::update_stats(size_t length) {
   image_ctx.perfcounter->inc(l_librbd_ws_bytes, length);
 }
 
+template <typename I>
+uint64_t ImageCompareAndWriteRequest<I>::append_journal_event(
+    const ObjectRequests &requests, bool synchronous) {
+
+  I &image_ctx = this->m_image_ctx;
+
+  uint64_t tid = 0;
+  uint64_t buffer_offset = 0;
+  assert(!this->m_image_extents.empty());
+  for (auto &extent : this->m_image_extents) {
+    bufferlist sub_bl;
+    sub_bl.substr_of(m_bl, buffer_offset, extent.second);
+    buffer_offset += extent.second;
+
+    tid = image_ctx.journal->append_compare_and_write_event(extent.first, extent.second,
+                                                m_cmp_bl, sub_bl, requests, synchronous);
+  }
+
+  if (image_ctx.object_cacher == NULL) {
```

Comment:
I think this should bypass the cache, so just always associate the completion w/ the journal event.